app/src/server/api/routers/datasetEntries.router.ts

```diff
@@ -260,20 +264,42 @@ export const datasetEntriesRouter = createTRPCRouter({
         ]);
       }
 
-      await prisma.datasetEntry.update({
+      const prevEntry = await prisma.datasetEntry.update({
         where: { id: input.id },
         data: {
-          type: input.updates.type,
-          input: parsedInput,
-          output: parsedOutput,
-          inputTokens,
-          outputTokens,
+          outdated: true,
+        },
+        include: {
+          matchedRules: {
+            select: {
+              pruningRuleId: true,
+            },
+          },
+        },
+      });
+
+      const newEntry = await prisma.datasetEntry.create({
+        data: {
+          input: parsedInput ?? prevEntry.input,
+          output: parsedOutput ?? prevEntry.output,
+          inputTokens: inputTokens ?? prevEntry.inputTokens,
+          outputTokens: outputTokens ?? prevEntry.outputTokens,
+          type: input.updates.type ?? prevEntry.type,
+          datasetId: prevEntry.datasetId,
+          sortKey: prevEntry.sortKey,
+          authoringUserId: ctx.session?.user.id,
+          persistentId: prevEntry.persistentId,
+          matchedRules: {
```

Comment:
I think we need to recalculate matched rules on every update?

Reply:
That's what we're doing on line 300 in `updatePruningRuleMatches`.